### DISCUSSION TITLE
Google Auth redirect issue using Vercel

### DISCUSSION THREAD

**Discussion by mrpbennett** [supabase/supabase#6832] [Questions]
Hi All,

I am having slight issues when using Google Auth redirect with my project. I have set the Google auth for internal folk only, this is how I have set up the URLs in supabase.

<img width="719" alt="Screenshot 2022-05-13 at 16 37 42" src="https://user-images.githubusercontent.com/1844080/168318919-4f285324-ec72-4ed7-a3c9-3090d4e5fb8e.png">

But when trigging the below function I just get redirected to `https://s3-directory-npi.vercel.app/home#access_token=xxx` with a 404

```javascript
const signInWithGoogle = async () => {
    const {user, session, error} = await supabase.auth.signIn(
      {
        provider: 'google',
      },
      {
        redirectTo: 'https://s3-directory-npi.vercel.app/home',
      },
    )

    if (error) {
      popupValidation('error', error.message)
    }
  }
```

I have also tried this 

<img width="797" alt="Screenshot 2022-05-13 at 16 48 47" src="https://user-images.githubusercontent.com/1844080/168320424-b0df5c5c-aa12-4398-a784-40f8ca9ef686.png">

Any suggestions to resolve the issue would be great.

**Comment by kangmingtay** [comment]
Hey @mrpbennett, it seems like you are being redirected to the redirect url you specified. If you're getting a 404 from that redirect url, it's most likely because you configured your api route incorrectly. 

I've tried visiting https://s3-directory-npi.vercel.app/home and it returns a 404 too
<img width="1474" alt="image" src="https://user-images.githubusercontent.com/28647601/168346590-23e08cae-4402-4a3a-bca7-5eab15866418.png">

**Reply by mrpbennett** [reply]
Hey @kangmingtay . This might should be because you're not in my GSuite ... I have set the google auth for internal users only.

My React router routes are:

```javascript
const root = ReactDOM.createRoot(document.getElementById('root'))
root.render(
  <BrowserRouter>
    <Routes>
      <Route path="/" element={<App />} />
      <Route path="/home" element={<Home />} />
    </Routes>
  </BrowserRouter>
)
```

So I can't really see why this isn't working. This also works in a local environment, it's just when I start using Vercel it goes tits up

**Reply by kangmingtay** [reply]
have you checked this yet? (https://vercel.com/docs/concepts/edge-network/frequently-asked-questions#why-do-i-see-a-404-error-when-accessing-my-deployment)

**Reply by mrpbennett** [reply]
I have.

but this confuses me. Since I have built standard sites with many routes and hosted them on vercel. I haven’t had any issues with building them.

I can’t figure out what is different here.

**Comment by Jittojoyes98** [comment]
Hey i have also issues when deloying a app that uses supabase auth using vercel. Can anyone please help?

**Reply by kangmingtay** [reply]
can you please provide more information about your issues? or if it requires sending confidential information, feel free to reach out to us at https://app.supabase.com/support/new

**Comment by kiroain** [comment]
I am also having the same issue when I use Vercel. With Netlify, Supabase's Google OAuth Provider works well, but with Vercel, it doesn't.

**Reply by kangmingtay** [reply]
can you please provide more information about your issues? or if it requires sending confidential information, feel free to reach out to us at https://supabase.com/dashboard/support/new

**Comment by nicitaacom** [comment]
I get this redirect issue as well
I described it here - https://github.com/orgs/supabase/discussions/18751#discussioncomment-7488122


To reproduce (for full version check link above)
1.
![supabase scrennshot](https://user-images.githubusercontent.com/39565703/280708519-20ce791a-244e-4f07-bd5b-d217c559a6f5.png)
2. 'Continue with google' button that user click to login with google
```ts
  const { error } = await supabaseClient.auth.signInWithOAuth({
        provider: "google",
        options: { redirectTo: `${location.origin}/auth/callback/` },
      })
```
3.`/auth/callback/` route (redirect to `auth/client` route with query params)
```ts
  return NextResponse.redirect(
    `${requestUrl.origin}/auth/client?userId=${user?.id}&username=${user?.user_metadata.name}
    &email=${user?.user_metadata.email}&profile_picture_url=${user?.user_metadata.picture}`,
  )
```
4. `auth/client` route
```ts
useEffect(() => {
    userStore.setUser(userId ?? "", username ?? "", email ?? "", profile_picture_url ?? "")
    router.push("/")
    //to prevent error about too many re-renders
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
```

But on second step it supabase redirect me back to localhost:3000 in production (because I define it on screenhost) - if I change (on scheenshot url) localhost:3000 to production url I will be redirected to production (always - in development as well)

Try it out - https://23-store.vercel.app/

**Reply by AITRUPEACE** [reply]
@nicitaacom were you able to figure it out? I see your site's authentication works now. My issue has been getting my Supabase ENVIRONMENT VARIABLES to work (url and secret).

**Reply by nicitaacom** [reply]
> @nicitaacom were you able to figure it out? I see your site's authentication works now. My issue has been getting my Supabase ENVIRONMENT VARIABLES to work (url and secret).

I don't know what do you mean by it

If you mean did I setup redirection
on prod - to prod url
on dev - to dev url

then yes

**Comment by RedDotz20** [comment]
After deploying, I set the `NEXT_PUBLIC_VERCEL_URL` environment variable to the production URL directly. Additionally, I configure this URL within the `"SITE URL"` field under Authentication > URL Configuration in Supabase.

_**Vercel Preview URL docs (Supabase):**_
https://supabase.com/docs/guides/auth/redirect-urls#vercel-preview-urls

I also reference this as a guide:
[Supabase Auth with Next 14 (app router) redirectTo link always redirects to base path](https://stackoverflow.com/questions/78220204/supabase-auth-with-next-14-app-router-redirectto-link-always-redirects-to-base)

```typescript
'use server';

import { createClient } from '@/utils/supabase/server';
import { redirect } from 'next/navigation';

// dynamically get the correct URL depending on the environment
const getURL = () => {
	let url =
		process?.env?.NEXT_PUBLIC_SITE_URL ?? // Set this to your site URL in production env
		process?.env?.NEXT_PUBLIC_VERCEL_URL ?? // Automatically set by Vercel
		'http://localhost:3000/'; // Default to localhost during development

	// Ensure the URL starts with 'https://' unless it's localhost
	url = url.startsWith('http') ? url : `https://${url}`;
	// Ensure there's a trailing slash at the end of the URL
	url = url.endsWith('/') ? url : `${url}/`;

	return url;
};

export const signInWithGoogle = () => {
	const supabase = createClient();
	const redirectToURL = `${getURL()}auth/callback`; // Use dynamic base URL for redirection

	const { data, error } = await supabase.auth.signInWithOAuth({
		provider: 'google',
		options: {
			queryParams: {
				access_type: 'offline',
				prompt: 'consent',
			},
			redirectTo: redirectToURL,
		},
	});

	if (error) {
		console.log(error);
		redirect('/error');
	}

	if (data.url) {
		redirect(data.url);
	}
};

```

**Please note:** In my case, I configure the redirect URL as /auth/callback to ensure direct access to this route. Within that route, I then set the next target path, such as `/home` or any preferred destination for redirection. [supabase.com/docs/guides/auth/social-login/auth-google](https://supabase.com/docs/guides/auth/social-login/auth-google?queryGroups=environment&environment=server&queryGroups=framework&framework=nextjs#application-code)

```typescript
// auth/callback/route.ts
const next = searchParams.get('next') ?? '/home'; 
```

Hope it helps 😊

**Comment by voidao** [comment]
Try https://dev.to/david_bilsonn/how-to-fix-vercel-404-error-child-url-path-issues-n0o
